### Import required libraries

In [27]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
import lightgbm as lgb
from catboost import CatBoostClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, precision_score  
from sklearn.metrics import recall_score, f1_score, roc_auc_score
import warnings
warnings.filterwarnings('ignore')

In [2]:
# read data
df = pd.read_csv('/home/saif/Desktop/Customer Churn Analysis and Prediction/top_features.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7010 entries, 0 to 7009
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   TotalCharges      7010 non-null   float64
 1   MonthlyCharges    7010 non-null   float64
 2   tenure            7010 non-null   int64  
 3   Contract          7010 non-null   int64  
 4   PaymentMethod     7010 non-null   int64  
 5   InternetService   7010 non-null   int64  
 6   OnlineBackup      7010 non-null   int64  
 7   OnlineSecurity    7010 non-null   int64  
 8   gender            7010 non-null   int64  
 9   TechSupport       7010 non-null   int64  
 10  PaperlessBilling  7010 non-null   int64  
 11  MultipleLines     7010 non-null   int64  
 12  Churn             7010 non-null   int64  
dtypes: float64(2), int64(11)
memory usage: 712.1 KB


In [4]:
df.describe()

TotalCharges  MonthlyCharges       tenure     Contract  PaymentMethod  \
count   7010.000000     7010.000000  7010.000000  7010.000000    7010.000000   
mean    2290.353388       64.888666    32.520399     8.838374       2.317404   
std     2266.820832       30.064769    24.520441     9.546590       1.150581   
min       18.800000       18.250000     1.000000     1.000000       1.000000   
25%      408.312500       35.750000     9.000000     1.000000       1.000000   
50%     1403.875000       70.400000    29.000000     1.000000       2.000000   
75%     3807.837500       89.900000    56.000000    12.000000       3.000000   
max     8684.800000      118.750000    72.000000    24.000000       4.000000   

       InternetService  OnlineBackup  OnlineSecurity       gender  \
count      7010.000000   7010.000000     7010.000000  7010.000000   
mean          1.225963      0.131098        0.072611     0.504280   
std           0.777600      0.737334        0.705040     0.500017   
min           0.000000     -1.000000       -1.000000     0.000000   
25%           1.000000      0.000000        0.000000     0.000000   
50%           1.000000      0.000000        0.000000     1.000000   
75%           2.000000      1.000000        1.000000     1.000000   
max           2.000000      1.000000        1.000000     1.000000   

       TechSupport  PaperlessBilling  MultipleLines        Churn  
count  7010.000000       7010.000000    7010.000000  7010.000000  
mean      0.076177          0.593153       0.326248     0.264907  
std       0.707190          0.491281       0.643333     0.441315  
min      -1.000000          0.000000      -1.000000     0.000000  
25%       0.000000          0.000000       0.000000     0.000000  
50%       0.000000          1.000000       0.000000     0.000000  
75%       1.000000          1.000000       1.000000     1.000000  
max       1.000000          1.000000       1.000000     1.000000

In [5]:
df.columns

Index(['TotalCharges', 'MonthlyCharges', 'tenure', 'Contract', 'PaymentMethod',
       'InternetService', 'OnlineBackup', 'OnlineSecurity', 'gender',
       'TechSupport', 'PaperlessBilling', 'MultipleLines', 'Churn'],
      dtype='object')

In [6]:
df.sample(5)

TotalCharges  MonthlyCharges  tenure  Contract  PaymentMethod  \
2766        185.40           20.00      10         1              1   
1684        947.75           25.25      39        12              4   
4169       6590.50          104.05      62        24              1   
1137        775.30           75.20      11         1              4   
4051       5515.80           79.60      68        24              3   

      InternetService  OnlineBackup  OnlineSecurity  gender  TechSupport  \
2766                0            -1              -1       1           -1   
1684                1             0               0       1            0   
4169                2             0               0       1            1   
1137                2             0               0       1            0   
4051                1             1               1       0            1   

      PaperlessBilling  MultipleLines  Churn  
2766                 0              0      0  
1684                 1             -1      0  
4169                 1              1      0  
1137                 1              1      0  
4051                 1              0      0

## Applying min max scaling on TotalCharges, MonthlyCharges,tenure, and Contract columns

In [7]:
# apply min max scaling on TotalCharges, MonthlyCharges, and tenure

scaler = MinMaxScaler()
df['TotalCharges'] = scaler.fit_transform(df[['TotalCharges']])
df['MonthlyCharges'] = scaler.fit_transform(df[['MonthlyCharges']])
df['tenure'] = scaler.fit_transform(df[['tenure']])
df['Contract'] = scaler.fit_transform(df[['Contract']])

In [8]:
df.sample(5)

TotalCharges  MonthlyCharges    tenure  Contract  PaymentMethod  \
6362      0.700958        0.649751  1.000000       1.0              4   
3351      0.039159        0.013930  0.253521       1.0              2   
194       0.135460        0.010448  0.816901       1.0              4   
311       0.215492        0.704478  0.295775       0.0              1   
2600      0.128375        0.117910  0.535211       0.0              4   

      InternetService  OnlineBackup  OnlineSecurity  gender  TechSupport  \
6362                1             0               1       1            1   
3351                0            -1              -1       1           -1   
194                 0            -1              -1       1           -1   
311                 2             0               1       0            1   
2600                1             0               0       0            0   

      PaperlessBilling  MultipleLines  Churn  
6362                 1              1      0  
3351                 0              0      0  
194                  0              0      0  
311                  1              0      0  
2600                 0             -1      1

# Task 2 : Data Split

In [9]:
# split data into train and test
X = df.drop('Churn', axis=1)
y = df['Churn']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [10]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((5608, 12), (1402, 12), (5608,), (1402,))

# Task 4 : Model Selection
I will train these model in order from 1 to 7 

**1.Logistic Regression (Baseline Model)**

**2.Decision Tree Classifier**

**3.Random Forest Classifier**

**4.XGBoost**

**5.LightGBM**

**6.CatBoost**

**7.Support Vector Machine (SVM)**

### Logistic regression

In [12]:
# training logistic regression
lr = LogisticRegression()
lr.fit(X_train, y_train)

LogisticRegression()

In [13]:
# predict on test data
y_pred_lr = lr.predict(X_test)

In [16]:
# evaluate model accuracy,precision, recall, F1-score, and ROC-AUC.
lr_accuracy = accuracy_score(y_test, y_pred_lr)
lr_precision = precision_score(y_test, y_pred_lr)
lr_recall = recall_score(y_test, y_pred_lr)
lr_f1 = f1_score(y_test, y_pred_lr)
lr_roc_auc = roc_auc_score(y_test, y_pred_lr)
print('Accuracy:', lr_accuracy)
print('Precision:', lr_precision)
print('Recall:', lr_recall)
print('F1-score:', lr_f1)
print('ROC-AUC:', lr_roc_auc)


Accuracy: 0.8259629101283881
Precision: 0.6410256410256411
Recall: 0.5451713395638629
F1-score: 0.5892255892255891
ROC-AUC: 0.7272572701519592


**1.Accuracy (82.6%)** → The model is mostly correct, but that doesn’t mean it’s perfect.

**2.Precision (64.1%)** → When the model predicts a customer will leave, it’s right 64% of the time.

**3.Recall (54.5%)** → The model is missing almost **half** of the actual churners, which is a problem if you want to save them.

**4.F1-Score (58.9%)** → A balance between catching churners and being correct when predicting churn. Not great, but not terrible.

**5.ROC-AUC (72.7%)** → The model is better than random guessing but needs improvement.

**Key Takeaway:**
- it fails to catch **many real churners**. The goal is to stop customers from leaving,so  model should improve recall.

### Decision Tree Classifier

In [18]:
# Training Decision Tree Classifier

dt = DecisionTreeClassifier()
dt.fit(X_train, y_train)

DecisionTreeClassifier()

In [19]:
# prediction on test data
y_pred_dt = dt.predict(X_test)

In [20]:
# evaluate model accuracy,precision, recall, F1-score, and ROC-AUC.
dt_accuracy = accuracy_score(y_test, y_pred_dt)
dt_precision = precision_score(y_test, y_pred_dt)
dt_recall = recall_score(y_test, y_pred_dt)
dt_f1 = f1_score(y_test, y_pred_dt)
dt_roc_auc = roc_auc_score(y_test, y_pred_dt)
print('Accuracy:', dt_accuracy)
print('Precision:', dt_precision)
print('Recall:', dt_recall)
print('F1-score:', dt_f1)
print('ROC-AUC:', dt_roc_auc)

Accuracy: 0.7425106990014265
Precision: 0.4470899470899471
Recall: 0.5264797507788161
F1-score: 0.4835479256080114
ROC-AUC: 0.6665701251581407


**Logistic Regression model is better than the Decision Tree in almost every way**. Here's why:

**1.Logistic Regression is more accurate (82.6% vs. 74.3%)**, meaning it makes fewer mistakes overall.

**2.It gives better precision (64.1% vs. 44.7%)**, so when it predicts a customer will leave, it's right more often.

**3.Both models have similar recall (~54% vs. ~52%)**, meaning they catch about the same number of actual churners.

**4.Logistic Regression has a better balance (F1-score 58.9% vs. 48.4%)**, so it’s more reliable.

**5.It also does a better job of separating churners from non-churners (ROC-AUC 72.7% vs. 66.7%)**.

#### What This Means
Logistic Regression is the better model right now because it makes fewer mistakes and is more reliable.

# Random Forest Classifier

In [12]:
# Training Random Forest Classifier
rf = RandomForestClassifier()
rf.fit(X_train, y_train)

RandomForestClassifier()

In [13]:
#Prediction on test data
y_pred_rf = rf.predict(X_test)

In [14]:
# evaluate model accuracy,precision, recall, F1-score, and ROC-AUC.
rf_accuracy = accuracy_score(y_test, y_pred_rf)
rf_precision = precision_score(y_test, y_pred_rf)
rf_recall = recall_score(y_test, y_pred_rf)
rf_f1 = f1_score(y_test, y_pred_rf)
rf_roc_auc = roc_auc_score(y_test, y_pred_rf)
print('Accuracy:', rf_accuracy)
print('Precision:', rf_precision)
print('Recall:', rf_recall)
print('F1-score:', rf_f1)
print('ROC-AUC:', rf_roc_auc)

Accuracy: 0.8045649072753209
Precision: 0.5824561403508772
Recall: 0.5171339563862928
F1-score: 0.5478547854785478
ROC-AUC: 0.7035253500710371


Random Forest model is better than Decision Tree but still worse than Logistic Regression in some areas. Here’s a simple breakdown:

**1.Accuracy (80.5%)** → Better than Decision Tree (74.3%) but still lower than Logistic Regression (82.6%).

**2.Precision (58.2%)** → Better than Decision Tree (44.7%) but lower than Logistic Regression (64.1%).

**3.Recall (51.7%)** → Almost the same as before, meaning it’s catching about the same number of churners as Logistic Regression (~54.5%).

**4.F1-score (54.8%)** → Better than Decision Tree (48.4%) but still lower than Logistic Regression (58.9%).

**5.ROC-AUC (70.4%)** → Improved from Decision Tree (66.7%) but still lower than Logistic Regression (72.7%).

#### What This Means:

- Random Forest is a step up from Decision Tree but not as strong as Logistic Regression in this case.
- Precision dropped compared to Logistic Regression, meaning it’s making more false positive mistakes.

# XGBoost

In [16]:
# Training XGBoost Classifier
xgb = XGBClassifier()
xgb.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=None, ...)

In [17]:
# Prediction on test data
y_pred_xgb = xgb.predict(X_test)

In [18]:
# evaluate model accuracy,precision, recall, F1-score, and ROC-AUC.
xgb_accuracy = accuracy_score(y_test, y_pred_xgb)
xgb_precision = precision_score(y_test, y_pred_xgb)
xgb_recall = recall_score(y_test, y_pred_xgb)
xgb_f1 = f1_score(y_test, y_pred_xgb)
xgb_roc_auc = roc_auc_score(y_test, y_pred_xgb)
print('Accuracy:', xgb_accuracy)
print('Precision:', xgb_precision)
print('Recall:', xgb_recall)
print('F1-score:', xgb_f1)
print('ROC-AUC:', xgb_roc_auc)

Accuracy: 0.8074179743223966
Precision: 0.580952380952381
Recall: 0.5700934579439252
F1-score: 0.5754716981132076
ROC-AUC: 0.7239921498785306


XGBoost model is now performing better than Random Forest and Decision Tree, and it's almost as good as Logistic Regression! Here's a simple breakdown:

**1.Accuracy (80.7%)** → Slightly better than Random Forest (80.5%) but still lower than Logistic Regression (82.6%).

**2.Precision (58.1%)** → Similar to Random Forest (58.2%) but lower than Logistic Regression (64.1%).

**3.Recall (57.0%)** → Best recall so far! Better than Logistic Regression (54.5%) and Random Forest (51.7%).

**4.F1-score (57.5%)** → Higher than Random Forest (54.8%), but slightly lower than Logistic Regression (58.9%).

**5.ROC-AUC (72.4%)** → Almost equal to Logistic Regression (72.7%), meaning XGBoost is just as good at separating churners from non-churners.

#### What This Means:

- XGBoost is the best at catching churners (highest recall so far).
- It performs close to Logistic Regression overall.

# LightGBM

In [20]:
# Training lightGBM Classifier
lgb = lgb.LGBMClassifier()
lgb.fit(X_train, y_train)

[LightGBM] [Info] Number of positive: 1536, number of negative: 4072
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000211 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 609
[LightGBM] [Info] Number of data points in the train set: 5608, number of used features: 12
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.273894 -> initscore=-0.974953
[LightGBM] [Info] Start training from score -0.974953


LGBMClassifier()

In [21]:
# Prediction on test data
y_pred_lgb = lgb.predict(X_test)

In [22]:
# evaluate model accuracy,precision, recall, F1-score, and ROC-AUC.
lgb_accuracy = accuracy_score(y_test, y_pred_lgb)
lgb_precision = precision_score(y_test, y_pred_lgb)
lgb_recall = recall_score(y_test, y_pred_lgb)
lgb_f1 = f1_score(y_test, y_pred_lgb)
lgb_roc_auc = roc_auc_score(y_test, y_pred_lgb)
print('Accuracy:', lgb_accuracy)
print('Precision:', lgb_precision)
print('Recall:', lgb_recall)
print('F1-score:', lgb_f1)
print('ROC-AUC:', lgb_roc_auc)

Accuracy: 0.8102710413694721
Precision: 0.5925925925925926
Recall: 0.5482866043613707
F1-score: 0.5695792880258899
ROC-AUC: 0.7181766046783727


LightGBM model is performing slightly better than XGBoost in some areas but not in others. Here’s a simple comparison:

**1.Accuracy (81.0%)** → Slightly better than XGBoost (80.7%) but still lower than Logistic Regression (82.6%).

**2.Precision (59.3%)** → Better than XGBoost (58.1%) but still lower than Logistic Regression (64.1%).

**3.Recall (54.8%)** → Lower than XGBoost (57.0%) but similar to Logistic Regression (54.5%).

**4.F1-score (56.9%)** → Almost the same as XGBoost (57.5%), meaning both balance precision and recall well.

**5.ROC-AUC (71.8%)** → Slightly worse than XGBoost (72.4%) and Logistic Regression (72.7%), meaning it’s not the best at separating churners from non-churners.

#### What This Means:

- LightGBM and XGBoost are very close in performance.
- XGBoost is better at catching churners (higher recall).
- LightGBM is slightly more accurate and has better precision.
- Neither has beaten Logistic Regression yet.

# CatBoost

In [24]:
# Training catboost Classifier
cat = CatBoostClassifier()
cat.fit(X_train, y_train)

Learning rate set to 0.021512
0:	learn: 0.6776523	total: 59.1ms	remaining: 59.1s
1:	learn: 0.6633492	total: 70.1ms	remaining: 35s
2:	learn: 0.6507619	total: 73.1ms	remaining: 24.3s
3:	learn: 0.6381747	total: 76.8ms	remaining: 19.1s
4:	learn: 0.6271026	total: 86.5ms	remaining: 17.2s
5:	learn: 0.6162299	total: 89.5ms	remaining: 14.8s
6:	learn: 0.6060567	total: 93ms	remaining: 13.2s
7:	learn: 0.5964787	total: 96.5ms	remaining: 12s
8:	learn: 0.5866813	total: 100ms	remaining: 11s
9:	learn: 0.5781369	total: 103ms	remaining: 10.2s
10:	learn: 0.5700041	total: 107ms	remaining: 9.6s
11:	learn: 0.5621540	total: 123ms	remaining: 10.1s
12:	learn: 0.5555346	total: 125ms	remaining: 9.5s
13:	learn: 0.5486267	total: 132ms	remaining: 9.27s
14:	learn: 0.5418331	total: 135ms	remaining: 8.84s
15:	learn: 0.5357530	total: 138ms	remaining: 8.48s
16:	learn: 0.5309288	total: 141ms	remaining: 8.13s
17:	learn: 0.5251962	total: 146ms	remaining: 7.95s
18:	learn: 0.5199429	total: 156ms	remaining: 8.07s
19:	learn: 0.

In [25]:
#prediction on test data
y_pred_cat = cat.predict(X_test)

In [26]:
# evaluate model accuracy,precision, recall, F1-score, and ROC-AUC.
cat_accuracy = accuracy_score(y_test, y_pred_cat)
cat_precision = precision_score(y_test, y_pred_cat)
cat_recall = recall_score(y_test, y_pred_cat)
cat_f1 = f1_score(y_test, y_pred_cat)
cat_roc_auc = roc_auc_score(y_test, y_pred_cat)
print('Accuracy:', cat_accuracy)
print('Precision:', cat_precision)
print('Recall:', cat_recall)
print('F1-score:', cat_f1)
print('ROC-AUC:', cat_roc_auc)

Accuracy: 0.81169757489301
Precision: 0.5966101694915255
Recall: 0.5482866043613707
F1-score: 0.5714285714285714
ROC-AUC: 0.7191016740585762


CatBoost model is now the best overall among tree-based models, but it's still very close to LightGBM and XGBoost. Here's the breakdown:

**1.Accuracy (81.2%)** → Slightly better than LightGBM (81.0%) and XGBoost (80.7%), but still lower than Logistic Regression (82.6%).

**2.Precision (59.7%)** → Best among tree-based models! Higher than LightGBM (59.3%) and XGBoost (58.1%), though still lower than Logistic Regression (64.1%).

**3.Recall (54.8%)** → Same as LightGBM and slightly lower than XGBoost (57.0%).

**4.F1-score (57.1%)** → Slightly better than LightGBM (56.9%) but still lower than XGBoost (57.5%) and Logistic Regression (58.9%).

**5.ROC-AUC (71.9%)** → Almost identical to LightGBM (71.8%) and still lower than XGBoost (72.4%) and Logistic Regression (72.7%).

#### What This Means:

- CatBoost is the best overall tree-based model, slightly ahead of LightGBM and XGBoost.
- Still, Logistic Regression is the best model so far in terms of accuracy, precision, and ROC-AUC.
- XGBoost had the best recall (57.0%), meaning it was the best at catching churners.
- CatBoost has the best precision (59.7%) among tree models, meaning its churn predictions are more reliable.

# Support Vector Machine

In [ ]:
# Training SVM Classifier
svm = SVC()
svm.fit(X_train, y_train)

SVC()

In [29]:
# prediction on test data
y_pred_svm = svm.predict(X_test)

In [30]:
# evaluate model accuracy,precision, recall, F1-score, and ROC-AUC.
svm_accuracy = accuracy_score(y_test, y_pred_svm)
svm_precision = precision_score(y_test, y_pred_svm)
svm_recall = recall_score(y_test, y_pred_svm)
svm_f1 = f1_score(y_test, y_pred_svm)
svm_roc_auc = roc_auc_score(y_test, y_pred_svm)
print('Accuracy:', svm_accuracy)
print('Precision:', svm_precision)
print('Recall:', svm_recall)
print('F1-score:', svm_f1)
print('ROC-AUC:', svm_roc_auc)

Accuracy: 0.8259629101283881
Precision: 0.672645739910314
Recall: 0.4672897196261682
F1-score: 0.5514705882352942
ROC-AUC: 0.6998798274356558


Support Vector Machine (SVM) model performed differently compared to other models. Here’s how it compares:

#### SVM vs. Other Models

**1.Accuracy (82.6%)** → Same as Logistic Regression (Best so far!).

**2.Precision (67.3%)** → Highest among all models! Better than CatBoost (59.7%) and Logistic Regression (64.1%).

**3.Recall (46.7%)** → Much lower than XGBoost (57.0%) and CatBoost (54.8%), meaning it misses more real churners.

**4.F1-score (55.1%)** → Worse than Logistic Regression (58.9%), meaning it doesn't balance precision and recall as well.

**5.ROC-AUC (69.9%)** → Lower than Logistic Regression (72.7%) and XGBoost (72.4%), meaning it’s not as good at distinguishing churners from non-churners.

#### What This Means:
- SVM is as accurate as Logistic Regression, so it's very consistent.
- It has the best precision (67.3%), meaning its churn predictions are the most reliable.
- But recall is low (46.7%), meaning it misses many churners.
- Overall, it’s good if you care more about precise predictions rather than catching every churner.